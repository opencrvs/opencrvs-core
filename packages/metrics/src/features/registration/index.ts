import {
  DECLARATION_STATUS,
  DECLARATION_TYPE,
  USER_ACTION
} from '@metrics/features/registration/fhirUtils'

/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
export interface IBirthRegistrationFields extends IPoint {
  compositionId: string
  ageInDays: number | undefined
}

export interface IDeathRegistrationFields extends IPoint {
  compositionId: string
  ageInYears: number | undefined
  deathDays: number | undefined
}

export interface IUserAuditFields extends IPoint {
  data: string | undefined
  userAgent: string
  ipAddress: string
}

export interface IMarriageRegistrationFields extends IPoint {
  compositionId: string
  daysAfterEvent: number | undefined
}

export interface IPoint {
  time?: string
}

export type IAuthHeader =
  | {
      Authorization: string
      'x-correlation-id': string
    }
  | {
      Authorization: string
    }

export interface IBirthRegistrationTags {
  regStatus: string
  registrarPractitionerId: string
  practitionerRole: string
  eventLocationType: string
  gender: string | undefined
  timeLabel: string | undefined
  ageLabel: string | undefined
  dateLabel: string | undefined
  officeLocation?: string
}

export interface IMarriageRegistrationTags {
  regStatus: string
  timeLabel: string | undefined
  dateLabel: string | undefined
  registrarPractitionerId: string
  practitionerRole: string
  officeLocation?: string
}

export interface IUserAuditTags {
  action: string | USER_ACTION
  practitionerId: string | undefined
}

export interface IDeathRegistrationTags {
  regStatus: string
  registrarPractitionerId: string
  practitionerRole: string
  eventLocationType: string
  gender: string | undefined
  timeLabel: string | undefined
  ageLabel: string | undefined
  dateLabel: string | undefined
  mannerOfDeath: string
  causeOfDeath: string
  officeLocation?: string
}

export interface IInProgressDeclarationFields {
  compositionId: string
}

export interface IInProgressDeclarationTags {
  regStatus: string
  missingFieldSectionId: string
  missingFieldGroupId: string
  missingFieldId: string
  eventType: string
  officeLocation?: string
}
export interface ICertifiedTags {
  eventType: DECLARATION_TYPE
  officeLocation?: string
}

interface ILocationTags {
  officeLocation?: string
}

export interface ITimeLoggedFields {
  timeSpentEditing: number
  compositionId: string
}
export interface ICertifiedFields {
  compositionId: string
}

export interface ITimeLoggedTags {
  currentStatus: string
  trackingId: string
  eventType: string
  officeLocation?: string
  practitionerId?: string
}

export interface IDurationFields {
  durationInSeconds: number
  compositionId: string
  currentTaskId: string
  previousTaskId: string
}
export interface IDurationTags {
  currentStatus: string
  previousStatus: string
  eventType: string
}

export interface IDurationPoints {
  measurement: string
  tags: IDurationTags
  fields: IDurationFields
  timestamp: number | undefined
}
export interface ICertifiedPoints {
  measurement: string
  tags: ICertifiedTags
  fields: ICertifiedFields
  timestamp: number | undefined
}

export interface ITimeLoggedPoints {
  measurement: string
  tags: ITimeLoggedTags
  fields: ITimeLoggedFields
  timestamp: number | undefined
}

export interface IInProgressDeclarationPoints {
  measurement: string
  tags: IInProgressDeclarationTags
  fields: IInProgressDeclarationFields
  timestamp: number | undefined
}

export interface IDeathRegistrationPoints {
  measurement: string
  tags: IDeathRegistrationTags
  fields: IDeathRegistrationFields
  timestamp: number | undefined
}

export interface IBirthRegistrationPoints {
  measurement: string
  tags: IBirthRegistrationTags
  fields: IBirthRegistrationFields
  timestamp: number | undefined
}

export interface IUserAuditPoints {
  measurement: string
  tags: IUserAuditTags
  fields: IUserAuditFields
  timestamp: number | undefined
}

export interface IUserAuditBody {
  practitionerId?: string
  action: USER_ACTION | string
  additionalData?: Record<string, any>
}
export interface IUserModelData {
  _id: string
  username: string
  name: {
    use: string
    family: string
    given: string[]
  }[]
  email: string
  emailForNotification?: string
  mobile?: string
  status: string
  role: string
  creationDate?: string
  practitionerId: string
  primaryOfficeId: string
  device: string
}
export interface IPaymentPoints {
  measurement: string
  tags: ILocationTags & {
    eventType: DECLARATION_TYPE
    paymentType: 'certification' | 'correction'
  }
  fields: IPaymentFields
  timestamp: number | undefined
}
export interface ICorrectionPoint {
  measurement: string
  tags: ILocationTags & {
    eventType: DECLARATION_TYPE
    // CLERICAL_ERROR, MATERIAL_ERROR, MATERIAL_OMISSION, JUDICIAL_ORDER, OTHER
    reason: string
  }
  fields: { compositionId: string }
  timestamp: number | undefined
}

export interface IPaymentFields {
  total: number
  compositionId: string
}

export interface IDeclarationsStartedPoints {
  measurement: string
  tags: ILocationTags
  fields: IDeclarationsStartedFields
  timestamp: number | undefined
}

export interface IDeclarationsStartedFields {
  role: string
  status?: DECLARATION_STATUS | null
  compositionId: string
}

export interface IRejectedFields {
  compositionId: string
}

export interface IRejectedPoints {
  measurement: string
  tags: ILocationTags
  fields: IRejectedFields
  timestamp: number | undefined
}

export interface IAdvancedSearchFields {
  clientId: string
}

export interface IAdvancedSearchTags {
  ipAddress: string
}

export interface IAdvancedSearchPoints {
  measurement: string
  tags: IAdvancedSearchTags
  fields: IAdvancedSearchFields
  timestamp: number | undefined
}

export type IPoints =
  | IDurationPoints
  | ICertifiedPoints
  | ITimeLoggedPoints
  | IInProgressDeclarationPoints
  | IPaymentPoints
  | IBirthRegistrationPoints
  | IDeathRegistrationPoints
  | IDeclarationsStartedPoints
  | IRejectedPoints
  | IUserAuditPoints
  | IAdvancedSearchPoints
