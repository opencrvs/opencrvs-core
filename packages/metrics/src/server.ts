/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import * as Hapi from '@hapi/hapi'
import {
  HOST,
  PORT,
  CERT_PUBLIC_KEY_PATH,
  DEFAULT_TIMEOUT
} from '@metrics/constants'
import getPlugins from '@metrics/config/plugins'
import { getRoutes } from '@metrics/config/routes'
import { readFileSync } from 'fs'
import { influx } from '@metrics/influxdb/client'
import {
  INFLUX_DB,
  INFLUX_HOST,
  INFLUX_PORT
} from '@metrics/influxdb/constants'
import * as database from '@metrics/config/database'
import * as mongoDirect from '@metrics/config/hearthClient'

const publicCert = readFileSync(CERT_PUBLIC_KEY_PATH)

export async function createServer() {
  const server = new Hapi.Server({
    host: HOST,
    port: PORT,
    routes: {
      cors: { origin: ['*'] },
      payload: { maxBytes: 52428800, timeout: DEFAULT_TIMEOUT }
    }
  })

  await server.register(getPlugins())

  server.auth.strategy('jwt', 'jwt', {
    key: publicCert,
    verifyOptions: {
      algorithms: ['RS256'],
      issuer: 'opencrvs:auth-service',
      audience: 'opencrvs:metrics-user'
    },
    validate: () => ({
      isValid: true
    })
  })

  server.auth.default('jwt')

  const routes = getRoutes()
  server.route(routes)

  server.ext({
    type: 'onRequest',
    method(request: Hapi.Request & { sentryScope?: any }, h) {
      if (request.payload) {
        request.sentryScope?.setExtra('payload', request.payload)
      }
      return h.continue
    }
  })

  async function start() {
    return influx
      .getDatabaseNames()
      .then((names: any) => {
        if (!names.includes(INFLUX_DB)) {
          return influx.createDatabase(INFLUX_DB)
        }
      })
      .then(async () => {
        server.log('info', `InfluxDB started on ${INFLUX_HOST}:${INFLUX_PORT}`)
        await server.start()
        await database.start()
        await mongoDirect.start()
        server.log('info', `Metrics server started on ${HOST}:${PORT}`)
      })
      .catch((err: Error) => {
        server.log('info', `Error creating Influx database! ${err.stack}`)
        throw err
      })
  }

  async function stop() {
    await server.stop()
    await database.stop()
    await mongoDirect.stop()
    server.log('info', 'Metrics server stopped')
  }

  return { server, start, stop }
}
