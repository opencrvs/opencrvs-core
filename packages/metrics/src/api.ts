/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import fetch from 'node-fetch'
import { IAuthHeader } from '@metrics/features/registration'
import {
  fhirUrl,
  COUNTRY_CONFIG_URL,
  SEARCH_URL,
  USER_MANAGEMENT_URL,
  DOCUMENTS_URL
} from '@metrics/constants'
import { logger } from '@opencrvs/commons'

export function fetchFHIR<T = any>(
  suffix: string,
  authHeader: IAuthHeader,
  method = 'GET',
  body?: string
) {
  const url = [fhirUrl.replace(/\/$/, ''), suffix].join('/')
  return fetch(url, {
    method,
    headers: {
      'Content-Type': 'application/fhir+json',
      ...authHeader
    },
    body
  })
    .then((response) => {
      return response.json() as Promise<T>
    })
    .catch((error) => {
      return Promise.reject(new Error(`FHIR request failed: ${error.message}`))
    })
}

export function fetchTaskHistory(taskId: string, authHeader: IAuthHeader) {
  // querying task history in descending order in case task history
  // gets very big and we always need the latest history
  return fetchFHIR<fhir.Bundle>(
    `Task/${taskId}/_history?_sort=-meta.lastUpdated`,
    authHeader
  )
}

export const fetchLocation = async (
  locationId: string,
  authHeader: IAuthHeader
) => {
  return await fetchFHIR<fhir.Location & { id: string }>(
    locationId.startsWith('Location/') ? locationId : `Location/${locationId}`,
    authHeader
  )
}

export async function fetchLocationsByType(
  type: string,
  authHeader: IAuthHeader
): Promise<(fhir.Location & { id: string })[]> {
  const bundle = await fetchFHIR(`Location?_count=0&type=${type}`, authHeader)
  return bundle?.entry?.map((entry: fhir.BundleEntry) => entry.resource) ?? []
}

export async function fetchTaskIdByCompositionID(
  compositionId: string,
  authHeader: IAuthHeader
) {
  const taskBundle = await fetchFHIR(
    `Task?focus=Composition/${compositionId}`,
    authHeader
  )
  if (!taskBundle.entry[0] || !taskBundle.entry[0].resource) {
    return null
  }
  return taskBundle.entry[0].resource.id
}

export async function totalOfficesInCountry(authHeader: IAuthHeader) {
  const bundle: fhir.Bundle = await fetchFHIR(
    'Location?type=CRVS_OFFICE',
    authHeader
  )
  return bundle.total ?? 0
}

export async function fetchChildLocationsByParentId(
  locationId: string,
  authHeader: IAuthHeader
): Promise<fhir.Location[]> {
  const bundle = await fetchFHIR(
    `Location?_count=0&type=ADMIN_STRUCTURE&partof=${locationId}`,
    authHeader
  )
  return bundle?.entry?.map((entry: fhir.BundleEntry) => entry.resource) ?? []
}

export async function fetchChildLocationsWithTypeByParentId(
  locationId: string,
  locationType: string,
  authHeader: IAuthHeader
): Promise<fhir.Location[]> {
  const bundle = await fetchFHIR(
    `Location?_count=0&type=${locationType}&partof=${locationId}`,
    authHeader
  )
  return bundle?.entry?.map((entry: fhir.BundleEntry) => entry.resource) ?? []
}

export function fetchFromResource(
  suffix: string,
  authHeader: IAuthHeader,
  method = 'GET',
  body?: string
) {
  const url = [COUNTRY_CONFIG_URL.replace(/\/$/, ''), suffix].join('/')
  return fetch(url, {
    method,
    headers: {
      ...authHeader
    },
    body
  })
    .then((response) => {
      return response.json()
    })
    .catch((error) => {
      return Promise.reject(
        new Error(`RESOURCE request failed: ${error.message}`)
      )
    })
}

export function fetchAllFromSearch(authHeader: IAuthHeader) {
  return fetch(`${SEARCH_URL}search/all`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...authHeader
    }
  })
    .then((response) => {
      return response.json()
    })
    .catch((error) => {
      return Promise.reject(
        new Error(`Search request failed: ${error.message}`)
      )
    })
}

export const fetchPractitionerRole = async (
  practitionerId: string,
  authHeader: IAuthHeader
) => {
  const roleBundle: fhir.Bundle = await fetchFHIR(
    `PractitionerRole?practitioner=${practitionerId}`,
    authHeader
  )
  const practitionerRole =
    roleBundle &&
    roleBundle.entry &&
    roleBundle.entry &&
    roleBundle.entry.length > 0 &&
    (roleBundle.entry[0].resource as fhir.PractitionerRole)

  const roleCode =
    practitionerRole &&
    practitionerRole.code &&
    practitionerRole.code.length > 0 &&
    practitionerRole.code[0].coding &&
    practitionerRole.code[0].coding[0].code

  if (roleCode) {
    return roleCode
  } else {
    return Promise.reject(new Error(`Role code cannot be found`))
  }
}

export interface ICountByLocation {
  total: number
  locationId: string
}

export async function countRegistrarsByLocation(
  authHeader: IAuthHeader,
  locationId?: string
): Promise<{ registrars: number }> {
  const res = await fetch(`${USER_MANAGEMENT_URL}/countUsersByLocation`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...authHeader
    },
    body: JSON.stringify({
      locationId
    })
  })
  return res.json()
}

export async function uploadFileToMinio(fileData: Buffer): Promise<string> {
  try {
    const result = await fetch(`${DOCUMENTS_URL}/upload-vs-export`, {
      method: 'post',
      headers: {
        'Content-Type': 'text/csv'
      },
      body: fileData
    })
    const res = await result.json()
    return res.refUrl
  } catch (err) {
    logger.error(err)
    return err
  }
}
