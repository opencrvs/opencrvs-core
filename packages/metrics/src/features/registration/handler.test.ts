/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createServer } from '@metrics/server'
import * as api from '@metrics/api'
import { readFileSync } from 'fs'
import * as jwt from 'jsonwebtoken'
import * as fetchAny from 'jest-fetch-mock'
import { testDeclaration } from '@metrics/features/registration/testUtils'
import { cloneDeep } from 'lodash'

jest.mock('@metrics/features/performance/viewRefresher', () => {
  const actualModule = jest.requireActual(
    '@metrics/features/performance/viewRefresher'
  )

  return {
    ...actualModule,
    refresh: jest.fn()
  }
})

const fetch = fetchAny as any
const fetchTaskHistory = api.fetchTaskHistory as jest.Mock

const token = jwt.sign(
  { scope: ['declare'] },
  readFileSync('./test/cert.key'),
  {
    algorithm: 'RS256',
    issuer: 'opencrvs:auth-service',
    audience: 'opencrvs:metrics-user'
  }
)

jest.mock('../metrics/utils', () => {
  const originalModule = jest.requireActual('../metrics//utils')
  return {
    __esModule: true,
    ...originalModule,
    getRegistrationTargetDays: () => 45
  }
})

describe('When a new registration event is received', () => {
  let server: any

  beforeEach(async () => {
    server = await createServer()
    fetch.mockResponses(
      [
        JSON.stringify({
          id: '1',
          partOf: {
            reference: 'Location/4'
          }
        })
      ],
      [
        JSON.stringify({
          id: '2',
          partOf: {
            reference: 'Location/3'
          }
        })
      ],
      [
        JSON.stringify({
          id: '3',
          partOf: {
            reference: 'Location/2'
          }
        })
      ]
    )
  })

  it('returns ok for valid new birth declaration', async () => {
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-notification-for-review',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload: testDeclaration
    })

    expect(res.statusCode).toBe(200)
  })

  it('returns 500 for invalid payload in new birth registration', async () => {
    const payload = cloneDeep(testDeclaration)
    // @ts-ignore
    payload.entry[0] = {}
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-notification-for-review',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })

    expect(res.statusCode).toBe(500)
  })
})

describe('When an existing declaration is marked registered', () => {
  let server: any

  beforeEach(async () => {
    server = await createServer()
  })
  it('writes the delta between DECLARED and VALIDATED states to influxdb', async () => {
    /* eslint-disable @typescript-eslint/no-require-imports */
    /* eslint-disable @typescript-eslint/no-var-requires */
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/sent-for-approval-request.json')
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-for-approval',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const declarationEventPoint =
      influxClient.writePoints.mock.calls[1][0].find(
        ({ measurement }: { measurement: string }) =>
          measurement === 'declaration_event_duration'
      )
    expect(res.statusCode).toBe(200)
    expect(declarationEventPoint).toMatchSnapshot()
  })
  it('writes the delta between DECLARED and REGISTERED states to influxdb', async () => {
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/mark-registered-request.json')
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/registered',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const declarationEventPoint =
      influxClient.writePoints.mock.calls[1][0].find(
        ({ measurement }: { measurement: string }) =>
          measurement === 'declaration_event_duration'
      )
    expect(res.statusCode).toBe(200)
    expect(declarationEventPoint).toMatchSnapshot()
  })
  it('writes the delta between VALIDATED and WAITING_VALIDATION states to influxdb', async () => {
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/waiting-validation-request.json')
    const taskHistory = require('./test-data/task-history-validated-response.json')
    fetchTaskHistory.mockReset()
    fetchTaskHistory.mockResolvedValue(taskHistory)
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/waiting-external-validation',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const declarationEventPoint =
      influxClient.writePoints.mock.calls[0][0].find(
        ({ measurement }: { measurement: string }) =>
          measurement === 'declaration_event_duration'
      )
    expect(res.statusCode).toBe(200)
    expect(declarationEventPoint).toMatchSnapshot()
  })

  it('writes the delta between VALIDATED and REGISTERED states to influxdb', async () => {
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/mark-registered-request.json')
    const taskHistory = require('./test-data/task-history-validated-response.json')
    fetchTaskHistory.mockReset()
    fetchTaskHistory.mockResolvedValue(taskHistory)
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/registered',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const declarationEventPoint =
      influxClient.writePoints.mock.calls[1][0].find(
        ({ measurement }: { measurement: string }) =>
          measurement === 'declaration_event_duration'
      )
    expect(res.statusCode).toBe(200)
    expect(declarationEventPoint).toMatchSnapshot()
  })

  describe('a death declaration', () => {
    it('writes the delta between REGISTERED and CERTIFIED states to influxdb', async () => {
      const influxClient = require('@metrics/influxdb/client')
      const payload = require('./test-data/mark-death-registered-request.json')
      const res = await server.server.inject({
        method: 'POST',
        url: '/events/death/registered',
        headers: {
          Authorization: `Bearer ${token}`
        },
        payload
      })
      const declarationEventPoint =
        influxClient.writePoints.mock.calls[1][0].find(
          ({ measurement }: { measurement: string }) =>
            measurement === 'declaration_event_duration'
        )

      expect(res.statusCode).toBe(200)
      expect(declarationEventPoint).toMatchSnapshot()
    })
  })
})
describe('When an existing declaration is marked certified', () => {
  let server: any

  beforeEach(async () => {
    server = await createServer()
  })

  it('writes the delta between REGISTERED and CERTIFIED states to influxdb', async () => {
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/mark-certified-request.json')
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/certified',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const declarationEventPoint =
      influxClient.writePoints.mock.calls[1][0].find(
        ({ measurement }: { measurement: string }) =>
          measurement === 'declaration_event_duration'
      )

    expect(res.statusCode).toBe(200)
    expect(declarationEventPoint).toMatchSnapshot()
  })
  describe('a death declaration', () => {
    it('writes the delta between REGISTERED and CERTIFIED states to influxdb', async () => {
      const influxClient = require('@metrics/influxdb/client')
      const payload = require('./test-data/mark-death-certified-request.json')
      const res = await server.server.inject({
        method: 'POST',
        url: '/events/death/certified',
        headers: {
          Authorization: `Bearer ${token}`
        },
        payload
      })
      expect(res.statusCode).toBe(200)
      const declarationEventPoint =
        influxClient.writePoints.mock.calls[1][0].find(
          ({ measurement }: { measurement: string }) =>
            measurement === 'declaration_event_duration'
        )

      expect(declarationEventPoint).toMatchSnapshot()
    })
  })
  describe('a birth declaration', () => {
    it('writes the declaration_event_duration to influxdb', async () => {
      const influxClient = require('@metrics/influxdb/client')
      const payload = require('./test-data/mark-certified-request.json')
      const res = await server.server.inject({
        method: 'POST',
        url: '/events/birth/certified',
        headers: {
          Authorization: `Bearer ${token}`
        },
        payload
      })
      expect(res.statusCode).toBe(200)
      const declarationEventDurationPoint =
        influxClient.writePoints.mock.calls[1][0].find(
          ({ measurement }: { measurement: string }) => {
            return measurement === 'declaration_event_duration'
          }
        )

      expect(declarationEventDurationPoint).toMatchSnapshot()
    })
  })
})

describe('When an in-progress declaration is received', () => {
  let server: any

  beforeEach(async () => {
    server = await createServer()
  })

  it('writes the in complete field points to influxdb', async () => {
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/sent-notification-request.json')
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-notification',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const inCompleteFieldPoints =
      influxClient.writePoints.mock.calls[1][0].find(
        ({ measurement }: { measurement: string }) =>
          measurement === 'in_complete_fields'
      )

    expect(res.statusCode).toBe(200)
    expect(inCompleteFieldPoints).toMatchSnapshot()
  })
  it('returns 500 for payload without expected extension on task resource', async () => {
    const payload = require('./test-data/sent-notification-request.json')
    payload.entry[1].resource.extension = []
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-notification',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    expect(res.statusCode).toBe(500)
  })

  it('writes the rejected points to influxdb', async () => {
    const influxClient = require('@metrics/influxdb/client')
    const payload = require('./test-data/sent-for-updates-request.json')
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-for-updates',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    const rejectedPoints = influxClient.writePoints.mock.calls[1][0].find(
      ({ measurement }: { measurement: string }) =>
        measurement === 'declarations_rejected'
    )

    expect(res.statusCode).toBe(200)
    expect(rejectedPoints).toMatchSnapshot()
  })
  it('returns 500 for payload without task resource', async () => {
    const payload = require('./test-data/sent-for-updates-request.json')
    payload.entry = []
    const res = await server.server.inject({
      method: 'POST',
      url: '/events/birth/sent-for-updates',
      headers: {
        Authorization: `Bearer ${token}`
      },
      payload
    })
    expect(res.statusCode).toBe(500)
  })
})
