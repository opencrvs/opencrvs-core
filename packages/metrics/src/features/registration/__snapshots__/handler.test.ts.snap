// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When an existing declaration is marked certified a birth declaration writes the declaration_event_duration to influxdb 1`] = `
{
  "fields": {
    "compositionId": "1744fc19-95be-434b-8bc9-a76d6b81551b",
    "currentTaskId": "dcf1644e-362d-414f-b0cc-2b240eea8591",
    "durationInSeconds": 94381,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "CERTIFIED",
    "eventType": "BIRTH",
    "previousStatus": "REGISTERED",
  },
  "timestamp": 1570613507666000000,
}
`;

exports[`When an existing declaration is marked certified a death declaration writes the delta between REGISTERED and CERTIFIED states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "62168b7e-7f80-4e2a-a7b0-1d33759c9d71",
    "currentTaskId": "75ba4bd8-009e-4a02-896c-148979dde09a",
    "durationInSeconds": 3564388,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "CERTIFIED",
    "eventType": "DEATH",
    "previousStatus": "REGISTERED",
  },
  "timestamp": 1574083515546000000,
}
`;

exports[`When an existing declaration is marked certified writes the delta between REGISTERED and CERTIFIED states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "1744fc19-95be-434b-8bc9-a76d6b81551b",
    "currentTaskId": "dcf1644e-362d-414f-b0cc-2b240eea8591",
    "durationInSeconds": 94381,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "CERTIFIED",
    "eventType": "BIRTH",
    "previousStatus": "REGISTERED",
  },
  "timestamp": 1570613507666000000,
}
`;

exports[`When an existing declaration is marked registered a death declaration writes the delta between REGISTERED and CERTIFIED states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "62168b7e-7f80-4e2a-a7b0-1d33759c9d71",
    "currentTaskId": "75ba4bd8-009e-4a02-896c-148979dde09a",
    "durationInSeconds": 3727875,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "REGISTERED",
    "eventType": "DEATH",
    "previousStatus": "DECLARED",
  },
  "timestamp": 1574074172990000000,
}
`;

exports[`When an existing declaration is marked registered writes the delta between DECLARED and REGISTERED states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "80433e78-7042-4837-b888-a79c0181bf4e",
    "currentTaskId": "18ea2464-f843-45e2-93b1-8b026e0d93a4",
    "durationInSeconds": 170630,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "REGISTERED",
    "eventType": "BIRTH",
    "previousStatus": "DECLARED",
  },
  "timestamp": 1570516927666000000,
}
`;

exports[`When an existing declaration is marked registered writes the delta between DECLARED and VALIDATED states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "80433e78-7042-4837-b888-a79c0181bf4e",
    "currentTaskId": "18ea2464-f843-45e2-93b1-8b026e0d93a4",
    "durationInSeconds": 170630,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "VALIDATED",
    "eventType": "BIRTH",
    "previousStatus": "DECLARED",
  },
  "timestamp": 1570516927666000000,
}
`;

exports[`When an existing declaration is marked registered writes the delta between VALIDATED and REGISTERED states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "80433e78-7042-4837-b888-a79c0181bf4e",
    "currentTaskId": "18ea2464-f843-45e2-93b1-8b026e0d93a4",
    "durationInSeconds": 170630,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "REGISTERED",
    "eventType": "BIRTH",
    "previousStatus": "VALIDATED",
  },
  "timestamp": 1570516927666000000,
}
`;

exports[`When an existing declaration is marked registered writes the delta between VALIDATED and WAITING_VALIDATION states to influxdb 1`] = `
{
  "fields": {
    "compositionId": "80433e78-7042-4837-b888-a79c0181bf4e",
    "currentTaskId": "18ea2464-f843-45e2-93b1-8b026e0d93a4",
    "durationInSeconds": 170630,
    "previousTaskId": "e172627c-8490-44da-a702-80a082f978f5",
  },
  "measurement": "declaration_event_duration",
  "tags": {
    "currentStatus": "WAITING_VALIDATION",
    "eventType": "BIRTH",
    "previousStatus": "VALIDATED",
  },
  "timestamp": 1570516927666000000,
}
`;

exports[`When an in-progress declaration is received writes the in complete field points to influxdb 1`] = `
{
  "fields": {
    "compositionId": "b2fbb82c-a68d-4793-98e1-87484fc785c4",
  },
  "measurement": "in_complete_fields",
  "tags": {
    "eventType": "BIRTH",
    "missingFieldGroupId": "child-view-group",
    "missingFieldId": "placeOfBirth",
    "missingFieldSectionId": "child",
    "officeLocation": "Location/b49503bf-531d-4642-ae1b-13f647b88ec6",
    "regStatus": "IN_PROGRESS",
  },
  "timestamp": 1551785895844000000,
}
`;

exports[`When an in-progress declaration is received writes the rejected points to influxdb 1`] = `
{
  "fields": {
    "compositionId": "81278acf-6105-435e-b1c2-91619c8cf6e1",
  },
  "measurement": "declarations_rejected",
  "tags": {
    "eventType": "BIRTH",
    "officeLocation": "Location/2a520dc1-0a9a-48a1-a4b8-66f3075a9155",
    "startedBy": "4bb03541-116e-4ce2-8ea4-407f307f7579",
  },
  "timestamp": 1589199329739000000,
}
`;
